File: OC/BaseVC/DCWKWebViewController.m
//
//  DCWKWebViewController.m
//  FYJ
//
//  Created by ma on 2018/8/31.
//  Copyright © 2018年 issuser. All rights reserved.
//

#import "DCWKWebViewController.h"
#import <WebKit/WebKit.h>
NSString *const huizhi = @"https://weixin.foyunjian.com/index.php/home/mine/custom/id/9";//慧值声明
NSString *const aboutUs = @"https://weixin.foyunjian.com/index.php/home/mine/custom/id/8";//关于我们
NSString *const buyMoney = @"https://weixin.foyunjian.com/index.php/home/mine/custom/id/7";//购币说明
@interface DCWKWebViewController ()<WKNavigationDelegate>
@property(nonatomic, strong) WKWebView *webView;
@property(nonatomic, strong) UIActivityIndicatorView *indicatorView;
@end

@implementation DCWKWebViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    _webView = [[WKWebView alloc] initWithFrame:CGRectMake(0, 0, SCREEN_WIDTH, SCREEN_HEIGHT-StatusBarAndNavigationBarHeight)];
    _webView.navigationDelegate = self;
    [self.view addSubview:_webView];
    NSURL *url = [NSURL URLWithString:[_urlString stringByAddingPercentEncodingWithAllowedCharacters:[NSCharacterSet URLFragmentAllowedCharacterSet]]];
    NSURLRequest *request = [NSURLRequest requestWithURL:url];
    [_webView loadRequest:request];
    _indicatorView = [[UIActivityIndicatorView alloc]initWithFrame:CGRectMake(0,0,50, 50)];
    _indicatorView.center = _webView.center;
    _indicatorView.activityIndicatorViewStyle = UIActivityIndicatorViewStyleGray;
    [self.view addSubview:_indicatorView];
    [_indicatorView startAnimating];
    if (!_wkTitle) {
        [self.webView addObserver:self forKeyPath:@"title" options:NSKeyValueObservingOptionNew context:NULL];
    }else {
        self.title = _wkTitle;
    }
    
}
#pragma mark - WKNavigationDelegate
- (void)webView:(WKWebView *)webView decidePolicyForNavigationAction:(WKNavigationAction *)navigationAction decisionHandler:(void (^)(WKNavigationActionPolicy))decisionHandler{
    if ([navigationAction.request.URL.absoluteString isEqual:@"about:blank"]){
        
        decisionHandler(WKNavigationActionPolicyCancel);
        
        return;
    }
   
    decisionHandler(WKNavigationActionPolicyAllow);
    
}
- (void)webView:(WKWebView *)webView didStartProvisionalNavigation:(WKNavigation *)navigation {
    
    
}

-(void)webView:(WKWebView *)webView didFinishNavigation:(WKNavigation *)navigation {
    [_indicatorView stopAnimating];
}
- (void)webView:(WKWebView *)webView didFailProvisionalNavigation:(WKNavigation *)navigation withError:(NSError *)error {
    
    [_indicatorView stopAnimating];
}
#pragma mark - kvo获取网页标题
- (void)observeValueForKeyPath:(NSString *)keyPath ofObject:(id)object change:(NSDictionary<NSKeyValueChangeKey,id> *)change context:(void *)context {
    if ([keyPath isEqualToString:@"title"]) {
        if (object == self.webView)
        {
            self.title = self.webView.title;
        }
        else {
            [super observeValueForKeyPath:keyPath ofObject:object change:change context:context];
        }
        
    }
}
- (void)dealloc{
    if (!_wkTitle) {
         [_webView removeObserver:self forKeyPath:@"title" context:nil];
    }
}
- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
}
@end

File: Swift/BaseVC/DCWKWebViewController.swift
//
//  DCWKWebViewController.swift
//  SwiftDemo
//
//  Created by ma on 2018/10/18.
//  Copyright © 2018年 ma. All rights reserved.
//

import UIKit
import WebKit
class DCWKWebViewController: DCBaseViewController {

    override func viewDidLoad() {
        super.viewDidLoad()
        self.webView = WKWebView.init(frame:self.view.bounds)
        self.webView?.navigationDelegate = self
        self.view.addSubview(self.webView!)
        if (urlString != nil) {
            let url = URL.init(string: self.urlString!)
            let request = URLRequest.init(url: url!)
            self.webView?.load(request)
            self.indicatorView = UIActivityIndicatorView.init(frame: CGRect.init(x: 0, y: 0, width: 50, height: 50))
            self.indicatorView?.center = self.webView!.center
            self.indicatorView?.style = UIActivityIndicatorView.Style.gray
            self.view.addSubview(self.indicatorView!)
            self.indicatorView?.startAnimating()
        }
        if (wkTitle == nil) {
            self.webView?.addObserver(self, forKeyPath: "title", options: NSKeyValueObservingOptions.new, context: nil)
        }else {
            self.title = self.wkTitle
        }
    
    }
     //MARK: - KVO监听标题
    override func observeValue(forKeyPath keyPath: String?, of object: Any?, change: [NSKeyValueChangeKey : Any]?, context: UnsafeMutableRawPointer?) {
        if keyPath == "title" {
            if (object as AnyObject) === self.webView {
                self.title = self.webView?.title
            }else {
                super.observeValue(forKeyPath: keyPath, of: object, change: change, context: context)
            }
        }
    }
    //MARK: - 私有属性
    private var webView:WKWebView?
    private var indicatorView:UIActivityIndicatorView?
    //MARK: - 公共属性
    var urlString:String?//{
//        didSet {
//            urlString = urlString?.addingPercentEncoding(withAllowedCharacters: CharacterSet.init())
//        }
   // }
    var wkTitle:String?
    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
    }
    deinit {
        if self.wkTitle != nil {
            self.webView?.removeObserver(self, forKeyPath: "title")
        }
    }
}
extension DCWKWebViewController:WKNavigationDelegate {
    func webView(_ webView: WKWebView, decidePolicyFor navigationAction: WKNavigationAction, decisionHandler: @escaping (WKNavigationActionPolicy) -> Void) {
        if navigationAction.request.url?.absoluteString == "about:blank" {
            decisionHandler(WKNavigationActionPolicy.cancel)
            return
        }
        decisionHandler(WKNavigationActionPolicy.allow)
        
    }
    func webView(_ webView: WKWebView, didFinish navigation: WKNavigation!) {
        self.indicatorView?.stopAnimating()
    }
    func webView(_ webView: WKWebView, didFail navigation: WKNavigation!, withError error: Error) {
        self.indicatorView?.stopAnimating()
    }
}
